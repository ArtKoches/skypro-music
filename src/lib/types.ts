export type tracksDataTypes = {
	_id: number
	name: string
	author: string
	release_date: string
	genre: string[]
	duration_in_seconds: number
	album: string
	logo: { type: string; data: unknown[] }
	track_file: string
	staredUser: number[]
}

export enum filterTypes {
	author = 'исполнителю',
	year = 'году выпуска',
	genre = 'жанру',
}

export type SignInDataType = {
	email: string
	password: string
}
