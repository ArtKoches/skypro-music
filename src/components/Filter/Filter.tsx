import React from 'react'
import styles from './Filter.module.css'
import classNames from 'classnames'

export default function Filter() {
	return (
		<div className={styles.centerblock__filter}>
			<div className={styles.filter__title}>Искать по:</div>
			<div className={classNames(styles.filter__button, styles._btn_text)}>
				исполнителю
			</div>
			<div className={classNames(styles.filter__button, styles._btn_text)}>
				году выпуска
			</div>
			<div className={classNames(styles.filter__button, styles._btn_text)}>
				жанру
			</div>
		</div>
	)
}
