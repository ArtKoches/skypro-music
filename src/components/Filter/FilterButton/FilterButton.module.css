.filterWrapper {
	position: relative;
}

.filter__button {
	font-style: normal;
	font-weight: 400;
	font-size: 16px;
	line-height: 24px;
	border: 1px solid #ffffff;
	border-radius: 60px;
	padding: 6px 20px;
}

.filter__button:not(:nth-child(2)) {
	margin-right: 10px;
}

._btn_text:hover {
	border-color: #d9b6ff;
	color: #d9b6ff;
	cursor: pointer;
}

._btn_text:active {
	border-color: #ad61ff;
	color: #ad61ff;
	cursor: pointer;
}
