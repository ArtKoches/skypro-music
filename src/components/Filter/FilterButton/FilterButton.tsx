import React from 'react'
import styles from './FilterButton.module.css'
import classNames from 'classnames'
import FilterList from '../FilterList/FilterList'

type Props = {
	title: string
	openFilterCategory: boolean
	handleOpenFilter: (title: string) => void
}

export default function FilterButton({
	title,
	openFilterCategory,
	handleOpenFilter,
}: Props) {
	return (
		<div className={styles.filterWrapper}>
			<div
				className={classNames(styles.filter__button, styles._btn_text)}
				onClick={() => handleOpenFilter(title)}
			>
				{title}
			</div>

			{openFilterCategory && <FilterList />}
		</div>
	)
}
