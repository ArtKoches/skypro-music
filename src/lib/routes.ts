export const routes = {
	HOME: '/',
	LOGIN: '/signin',
	REGISTER: '/signup',
}
