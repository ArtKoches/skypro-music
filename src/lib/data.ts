import { routes } from './routes'

export const selectionsList = [
	{
		id: 2,
		href: routes.SELECTION,
		src: '/img/playlist01.png',
		alt: 'days playlist',
	},
	{
		id: 3,
		href: routes.SELECTION,
		src: '/img/playlist02.png',
		alt: 'dance hits',
	},
	{
		id: 4,
		href: routes.SELECTION,
		src: '/img/playlist03.png',
		alt: 'indie charge',
	},
]
