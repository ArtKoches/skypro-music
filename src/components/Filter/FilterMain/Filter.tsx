'use client'

import React, { useState } from 'react'
import styles from './Filter.module.css'
import FilterButton from '../FilterButton/FilterButton'
import { filterTypes } from '@/lib/types'
import { useAppSelector } from '@/store/store'

const filterTitles: string[] = [
	filterTypes.author,
	filterTypes.year,
	filterTypes.genre,
]

export default function Filter() {
	const { currPlaylist } = useAppSelector(state => state.currPlaylist)
	const [openFilterCategory, setOpenFilterCategory] = useState<string | null>(
		null,
	)

	const handleOpenFilter = (title: string) => {
		if (openFilterCategory === title) {
			setOpenFilterCategory(null)
		} else {
			setOpenFilterCategory(title)
		}
	}

	const getUniqueFilterLists = (option: string): string[] => {
		switch (option) {
			case filterTypes.author:
				return Array.from(
					new Set<string>(currPlaylist.map(track => track.author)),
				)
			case filterTypes.year:
				return ['По умолчанию', 'Сначала новые', 'Сначала старые']
			case filterTypes.genre:
				return Array.from(
					new Set<string>(currPlaylist.map(track => track.genre).flat()),
				)
			default:
				return []
		}
	}

	return (
		<div className={styles.centerblock__filter}>
			<div className={styles.filter__title}>Искать по:</div>
			{filterTitles.map((title, key: number) => {
				const getFilterLists = getUniqueFilterLists(title)

				return (
					<FilterButton
						key={key}
						title={title}
						openFilterCategory={openFilterCategory === title}
						handleOpenFilter={handleOpenFilter}
						getFilterLists={getFilterLists}
					/>
				)
			})}
		</div>
	)
}
