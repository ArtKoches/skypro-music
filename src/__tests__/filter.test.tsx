import Filter from '@/components/Filter/FilterMain/Filter'
import renderer from 'react-test-renderer'
import { render, screen } from '@testing-library/react'
import FilterButton from '@/components/Filter/FilterButton/FilterButton'
import { FilterType } from '@/lib/types'

describe('Фильтр', () => {
	it('рендер заголовка', () => {
		render(<Filter playlist={[]} />)

		const filterTitle = screen.getAllByText('Искать по:')
		expect(filterTitle.length).toBeGreaterThan(0)
	})
	it('рендер всплывающего списка категории', () => {
		render(<Filter playlist={[]} />)

		const filterList = screen.getAllByText('исполнителю')
		expect(filterList.length).toBeGreaterThan(0)
	})

	const filterListTests: string[] = [
		FilterType.author,
		FilterType.year,
		FilterType.genre,
	]
	filterListTests.forEach(btnTitle => {
		it('рендер кнопок категории фильтров', () => {
			const filter = renderer
				.create(<FilterButton title={btnTitle} getFilterLists={[]} />)
				.toJSON()
			expect(filter).toMatchSnapshot()
		})
	})
})
