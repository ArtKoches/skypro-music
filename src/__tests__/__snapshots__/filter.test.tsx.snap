// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Фильтр рендер кнопок категории фильтров 1`] = `
<div
  className="filterWrapper"
>
  <div
    className="filter__button"
    onClick={[Function]}
  >
    исполнителю
  </div>
</div>
`;

exports[`Фильтр рендер кнопок категории фильтров 2`] = `
<div
  className="filterWrapper"
>
  <div
    className="filter__button"
    onClick={[Function]}
  >
    году выпуска
  </div>
</div>
`;

exports[`Фильтр рендер кнопок категории фильтров 3`] = `
<div
  className="filterWrapper"
>
  <div
    className="filter__button"
    onClick={[Function]}
  >
    жанру
  </div>
</div>
`;
